### ISSUE TITLE
[Spike] How to implement blend groups

### ISSUE THREAD

**Issue by RichardBrown384** [RichardBrown384/riscos-artworks-js#116]
# About

Blend groups are partially understood but not implemented. Implementing these would be a nice to have.

# Task

There are a number of issues left to investigate

- [ ] How does ArtWorks interpolate colours? Presumably not in the RGB colour space. Maybe HSV?
- [ ] How does ArtWorks interpolate the shapes? That is, what does it do in non-trivial cases when the number of vertices differ?
- [ ] How does ArtWorks interpolate the path attributes like stroke width, dash pattern, winding rule (!) etc
- [ ] How does ArtWorks deal with different fill types e.g. the start shape has a radial fill and the target has a linear fill.
- [ ] What happens when the colours aren't indexed into the palette (since experiments show that at least in trivial cases ArtWorks uses the colour model information in the the palette entries)?

# Potential useful links

[StackExchange question about Inkscape](https://graphicdesign.stackexchange.com/questions/127669/is-it-possible-to-do-smooth-color-transition-between-2-paths-in-inkscape)

**Comment by RichardBrown384** [comment]
General findings after #159

## Strokes
1. Stroke attributes aren't interpolated unless a join style is specified with the file.
2. Continuous stroke attributes are interpolated linearly.
3. Discrete stroke attributes (join style) switch over halfway through the interpolation.
4. Line caps and markers aren't interpolated, rather the paths are interpolated as closed.
5. Dash patterns are _bizarre_. Early experiments showed that the file must specify a dash pattern otherwise !AWViewer crashes but attempting to interpolate a dash pattern results in !AWViewer filling the viewport with the stroke colour. Further investigation required

## Fills
1. Flat to flat works as you would expect.
2. Flat to linear (and vice versa) works as you would expect
3. Flat to radial (and vice versa) works as you would expect
4. Linear to linear works as you would expect
5. Linear to radial (and vice versa) isn't fully supported and the rule appears to be that you you use one fill for half the interpolation steps and the other for the remaining ones.
6. Radial to radial doesn't work as expected. Further investigation required although the suspicion is that the end fill's start and end points are translated with the geometry.
7. Discrete fill attributes (such as winding order) switch over halfway through the interpolation.

Note there may be gaps as there are potentially render-relevant control words in the record headers that aren't fully understood and there maybe other factors influencing the rendering (e.g. up until the discovery that a join style was necessary for interpolated strokes to be rendered, it was thought that interpolated strokes weren't rendered at all.

**Comment by RichardBrown384** [comment]
General findings after #162 (Simple Geometry blends)

# Equal number of points

This behaves as you would expect the points are simply interpolated. When a line is being blended with a Bezier, the line is converted to a Bezier during the intermediate steps. The control points are placed at 1/3 and 2/3s of the way along the line.


# Differing number of points

For poly line geometry (i.e. no Beziers) the blending algorithm behaves more-or-less how one would expect it to. The algorithm takes the shape with the least number of points and then adds additional points so that the shapes can be interpolated linearly. How those additional points are chosen is unknown. 

1. Experiments with regular polygons would initially indicate that some sort of projection is used. However, when experimenting with blending between a square being rotated at 10 degree increments and a triangle, one would expect the point at which the corresponding triangle edge is split to rotate with the square but this is not the case. The same edge is split at its midpoint.
2. Experiments with a non-convex shape seem to confirm the additional points hypothesis.

# Self intersecting Beziers

Self intersecting Beziers aren't simply interpolated. It appears !AWViewer checks for self intersecting geometry and manipulates the paths prior to interpolation.

**Comment by RichardBrown384** [comment]
Documented examples are being written up [here](../tree/main/docs/blend-groups).

**Comment by RichardBrown384** [comment]
General geometry observations:

1. The main task appears to be to come up with a correspondence between the points on the source and target paths
2. The first point of the source path appears to correspond to the first point of the target path
3. The order in which the points are defined in the source and target paths is more important than their spacial locations
4. Once the correspondence has been determined, points are inserted into both the source and target paths
5. Once the correspondence has been determined, the insertion points appear to be determined by the arc lengths of the points involved (there's less evidence for Béziers but it appears to be the case)
6. There don't seem to be any checks in place to enforce pleasing intermediate paths (e.g. the interpolated geometry can be made to self-intersect). This could be because ArtWorks itself might have done some initial work to make sure that the input paths for blending are well behaved and !AWViewer is expecting well behaved inputs and doesn't perform its own checks.
7. You can scale the source and target paths independently and the point correspondence doesn't appear to change.
